Introducing stm32 MDIO driver

spectrum70 opened the pull request:
Introducing stm32 MDIO driver, needed for all ethernet phy drivers on top of it, as actually needed from adin1100.

MDIO hw module is part of the ETH IP module in both stm32 h5/h7 families (and also likely f, not interested from this patch-set).
This patch-set abstracts-out the MDIO related registers control, in a separate MDIO driver, so that further PHY drivers can be implemented, and MDIO and PHY apis can be used.

Below a simple unit test, where i write and read registers, testing the contents with previously written values:
```
[00:00:00.001,000] <inf> phy_mii: PHY (0) ID 283BC81

[00:00:00.002,000] <inf> phy_mii: PHY (0) 2.4V mode supported
[00:00:00.002,000] <inf> phy_mii: PHY (0) 2.4V mode disabled
[00:00:00.002,000] <inf> phy_mii: PHY (0) Link is up
[00:00:00.002,000] <inf> phy_mii: PHY (0) Link speed 10 Mb, full duplex

[00:00:00.003,000] <err> net_if: Cannot join all nodes address ff02::1 for 1 (-115)
[00:00:00.003,000] <err> net_if: Cannot join solicit node address ff02::1:ff67:9f3 for 1 (-115)
*** Booting Zephyr OS build v3.6.0-2008-g7a2ec098c980 ***
[00:00:00.003,000] <inf> net_config: Initializing network
[00:00:00.003,000] <inf> net_config: Waiting interface 1 (0x2400138c) to be up...
[00:00:00.503,000] <inf> net_config: Interface 1 (0x2400138c) coming up
[00:00:00.503,000] <inf> net_config: IPv4 address: 192.168.5.4
[00:00:00.503,000] <dbg> net_stats_sample: main: C22 MDIO read ...
[00:00:00.503,000] <dbg> net_stats_sample: main: C22 MDIO read -ok-
[00:00:00.503,000] <dbg> net_stats_sample: main: C22 MDIO write ...
[00:00:00.503,000] <dbg> net_stats_sample: main: C22 MDIO write -ok-
[00:00:00.503,000] <dbg> net_stats_sample: main: C22 PHY read ...
[00:00:00.503,000] <dbg> net_stats_sample: main: C22 PHY read -ok-
[00:00:00.503,000] <dbg> net_stats_sample: main: C22 PHY write ...
[00:00:00.503,000] <dbg> net_stats_sample: main: C22 PHY write -ok-
[00:00:01.503,000] <dbg> net_stats_sample: main: Getting link state, should be up now ...
[00:00:01.503,000] <dbg> net_stats_sample: main: Link state ... UP
[00:00:01.503,000] <dbg> net_stats_sample: main: Test completed \o/
uart:~$ 

```

Waiting for comments.
Thanks

spectrum70 committed:
drivers: mdio: add stm32 mdio support

MDIO is part of the ETH IP, but some phy chip may need a
specific phy driver to set up certain vendor registers enabling
particular features.

Add support for stm32 mdio access.

Signed-off-by: Angelo Dureghello <adureghello@baylibre.com>

spectrum70 committed:
dts: arm: st: add mdio node for h5 and h7

Add mdio node for h5 and h7 series.

Since MDIO registers are part of the same ETH hw IP, keeping mdio
node just as a child of mac/eth, cannot see as appropriate to assign
an adddress to it.

Signed-off-by: Angelo Dureghello <adureghello@baylibre.com>

spectrum70 committed:
boards: st: add mdio and phy nodes

Add mdio and phy nodes for h5/h7 families.

Signed-off-by: Angelo Dureghello <adureghello@baylibre.com>

gramsay0 commented:
Do you need to remove all the phy related stuff from the STM32 Ethernet drivers?
Is it possible to extend the existing phy_adin2111.c to support adin1100 (it looks like Linux combines them https://github.com/torvalds/linux/blob/master/drivers/net/phy/adin1100.c#L344)

spectrum70 commented:
Hi @gramsay0 

thanks for the feedbacks,

> Do you need to remove all the phy related stuff from the STM32 Ethernet drivers? Is it possible to extend the existing phy_adin2111.c to support adin1100 (it looks like Linux combines them 
https://github.com/torvalds/linux/blob/master/drivers/net/phy/adin1100.c#L344)

Well, adin2111/1110 are spi-based only, adin1100 (and adin1300 that will follow) are rmii interface based, and industrial ethernet stuff. A PHY driver for them is asked by ADI, since generic-phy (that i tested) is not working properly for them.

> 
> edit: that is it seems like the phy would be managed by both the phy and Ethernet drivers (and without synchronisation)

Yes, correct, if it's an issue i can remove PHY HAL access for H5/H7 from  eth_stm32_hal.c in case, and use a MDIO API call in that place.

gramsay0 commented:
> Well, adin2111/1110 are spi-based only, adin1100 (and adin1300 that will follow) are rmii interface based, and industrial ethernet stuff. A PHY driver for them is asked by ADI, since generic-phy (that i tested) is not working properly for them.

phy_adin2111.c does not use SPI, it uses the MDIO API which could be implemented by any MDIO driver.
mdio_adin2111.c calls directly into eth_adin2111.c functions which use SPI.

So it should be possible to use phy_adin2111.c (possibly with extensions for adin1100) with your stm32 MDIO driver.

> Yes, correct, if it's an issue i can remove PHY HAL access for H5/H7 from eth_stm32_hal.c in case, and use a MDIO API call in that place.

Someone more familiar with this area would be best to give you advice on this

decsny commented on drivers/mdio/Kconfig.stm32_hal:
```diff
@@ -0,0 +1,11 @@
+config MDIO_ST_STM32_HAL
+	bool "STM32 MDIO controller driver"
+	default y
+	depends on SOC_SERIES_STM32H7X || SOC_SERIES_STM32H5X || ETH_STM32_HAL_API_V2
```
is the dependency on the SOC kconfigs required if it already depends on the devicetree having the node enabled? presumably, a correct devictree would not have this node if the SOC does not contain the IP, no?

spectrum70 replied:
soubds correct. fixed

decsny commented on drivers/mdio/mdio_stm32_hal.c:
```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2024 BayLibre, SAS
+ * Copyright (c) 2010-2024 Analog Devices Inc.
```
I am wondering where did these 2010 ADI copyrights come from since the mdio API in zephyr was added in 2021? therefore I can't imagine how did this code exist in 2010?

spectrum70 replied:
done, thanks

decsny commented on drivers/mdio/mdio_stm32_hal.c:
```diff
@@ -0,0 +1,119 @@
+
+	k_sem_take(&dev_data->sem, K_FOREVER);
+
+	ret = HAL_ETH_WritePHYRegister(hmdio, prtad, regad, data);
```
I noticed that the stm32 hal looks for definition of  ETH_MDIO_BUS_TIMEOUT to decide how long to block and I am wondering if it would be worth any value to provide this definition based on the value of a kconfig so that it is configurable

spectrum70 replied:
It is,

#define ETH_MDIO_BUS_TIMEOUT    1000U

So it's actually 1 second max. It's a max limit, so busy flag is cleared to 0 much before.  Seems not simple for me  to replace this value, since part of stm32cube module repo.

decsny replied:
you could make a kconfig and give zephyr_compile_definitions in the cmake based on that config, but if you don't think this is worth it then I don't think it is a big deal to skip it in this PR

decsny commented on drivers/mdio/mdio_stm32_hal.c:
```diff
@@ -0,0 +1,119 @@
+	if (ret != HAL_OK)
+		return -EIO;
+
+	*data = read & 0xffff;
```
use GENMASK macro in zephyr/sys/util.h preferably or FIELD_GET

spectrum70 replied:
thanks, done

decsny commented on drivers/ethernet/phy/Kconfig:
```diff
@@ -33,6 +33,16 @@ config PHY_GENERIC_MII
+	depends on DT_HAS_ADI_ADIN1100_PHY_ENABLED
+	depends on ETH_STM32_HAL
+	depends on MDIO
+	depends on MDIO_ST_STM32_HAL
```
this should not be dependent on the implementations being stm32 hal, in principle the phy driver should be able to fit with any mdio controller or ethernet device

spectrum70 replied:
done, thanks, removed both MDIO_ST_STM32_HAL and ETH_STM32_HAL, since stm32 mdio driver goes trough stm32 HAL.

decsny replied:
it seems they are still there, is it intentional

decsny commented on drivers/ethernet/phy/phy_adin1100.c:
the other reviewer is right that the most important consumer of the phy abstraction layer is the ethernet mac driver itself, and honestly it would not be a good idea to have a separate phy driver while having basically another phy driver embedded/hardcoded per se in the mac driver, they could conflict without knowing about each other... I havent looked at the stm32 driver but if it really does use the stm hal phy calls then it would probably be in your interest to convert it to using the zephyr phy api and go through this driver (or whatever phy driver, since that is the idea)

spectrum70 replied:
@decsny anbd all, thanks for all the feedbacks,

so, based on the above suggestions, i would go in this direction now for a v2:

1) see if i can use same phy driver for adin2111 and adin1100
2) on the stm32 eth driver, willuse the phy api to avoid conflicts
3) keeping/fixing all your suggested points above

Let me know if you see any issue.

Thanks a lot

decsny replied:
thats fine

spectrum70 commented:
@decsny 

As of now, cannot use adin2111/1110 phy driver for adin1100, since several accesses are Clause 45, and stm32 mdio cannot implement it (missing HAL functions). I pushed a PR to ST HAL for this.

https://github.com/STMicroelectronics/stm32h7xx_hal_driver/pull/52

So right now, only way for me to proceed is to keep in adin1100 phy driver. If ST HAL will be accepted, i can cleanup code removing the phy driver.

gramsay0 commented:
> As of now, cannot use adin2111/1110 phy driver for adin1100, since several accesses are Clause 45, and stm32 mdio cannot implement it (missing HAL functions). I pushed a PR to ST HAL for this.
> 
> [STMicroelectronics/stm32h7xx_hal_driver#52](https://github.com/STMicroelectronics/stm32h7xx_hal_driver/pull/52)
> 
> So right now, only way for me to proceed is to keep in adin1100 phy driver. If ST HAL will be accepted, i can cleanup code removing the phy driver.

Both adin2111 and adin1110 support clause 22 and 45. So writing an adin1110 phy driver just to support clause 22 (because the existing adin2111 uses clause 45) does not seem like a good path forward. IMO a Kconfig to enable C45 support would seem more appropriate.

Although adding clause 45 support to the STM HAL is a great alternative too 👍

spectrum70 commented:
@gramsay0 

thanks,

well, ADI naming is quite funny :) there are ADIN2111/1110 (same eth/mdio/phy driver,  spi based) and this new ADIN1100 that is rmii. It require right now a new driver since from there i can access MMD registers using the "bridge, indirect accerss" feature of it, simulating Clause 45, that cannot be implemented now in STM32 mdio. So for now keeping in adin1100-phy.

gramsay0 commented:
> well, ADI naming is quite funny :) there are ADIN2111/1110 (same eth/mdio/phy driver, spi based) and this new ADIN1100 that is rmii. It requires right now a new driver since from there i can access MMD registers using the "bridge, indirect accerss" feature of it, simulating Clause 45, that cannot be implemented now in STM32 mdio. So for now keeping in adin1100-phy.

If a new variant supports a new feature, then you can add that to the device tree, and check / make use of the new feature in the driver, no need to duplicate the other functionality.

> If ST guys will accept HAL changhes, i promise clean up and to remove the adin1100 phy driver.

👍

spectrum70 commented:
> > well, ADI naming is quite funny :) there are ADIN2111/1110 (same eth/mdio/phy driver, spi based) and this new ADIN1100 that is rmii. It requires right now a new driver since from there i can access MMD registers using the "bridge, indirect accerss" feature of it, simulating Clause 45, that cannot be implemented now in STM32 mdio. So for now keeping in adin1100-phy.
> 
> If a new variant supports a new feature, then you can add that to the device tree, and check / make use of the new feature in the driver, no need to duplicate the other functionality.
> 
> > If ST guys will accept HAL changhes, i promise clean up and to remove the adin1100 phy driver.
> 
> 👍

Ack, will try to keep adin1110 phy with some conditionals for rmii

spectrum70 committed:
drivers: eth: eth_stm32: use phy api support for h5/h7

Use PHY apis to access the PHY, to avoid any kind of collisions
with other tasks using the PHY apis.

Signed-off-by: Angelo Dureghello <adureghello@baylibre.com>

spectrum70 committed:
drivers: eth: phy: adin2111: add support for adin1100 phy

Add support for similar adin1100 phy, boath are 10Base-T1L,
only difference is that adin1100 connects through r/mii.

Signed-off-by: Angelo Dureghello <angelo@kernel-space.org>

spectrum70 commented:
Pushed a v2:
- removed specific adin1100 phy driver, used existing adin2111/1110
- added compatible adin1100, to allow adin2111-phy driver to configure appropriately
- fixed above points
- keeping mdio node as child of eth, since mdio registers are part of same ETH IP
- retested over different boards

spectrum70 commented:
Attaching tests done.
[test-zephyr-adin1100.txt](https://github.com/zephyrproject-rtos/zephyr/files/14959608/test-zephyr-adin1100.txt)

decsny reviewed (commented):
should the phy address come from DT now instead of kconfig? in the stm32 hal driver

spectrum70 commented:
> should the phy address come from DT now instead of kconfig? in the stm32 hal driver

Hi, @decsny 
actually i am introducing the change for h5/h7, that i can test here. So for F serie PHY_ADDR from Kconfing is still needed.

carlocaione commented:
@erwango can you take a look?

gramsay0 reviewed (commented):
Just some thoughts on this, not necessarily requiring changes

spectrum70 commented on drivers/ethernet/phy/phy_adin2111.c:
```diff
@@ -304,6 +364,70 @@ static int phy_adin2111_cfg_link(const struct device *dev,
+	old_state = data->state;
+	data->state.is_up = !!(bmsr & MII_BMSR_LINK_STATUS);
+
+	if (memcmp(&old_state, &data->state, sizeof(struct phy_link_state)) != 0) {
```
good point, fixed comparing members.

spectrum70 commented on drivers/ethernet/phy/phy_adin2111.c:
```diff
@@ -304,6 +364,70 @@ static int phy_adin2111_cfg_link(const struct device *dev,
+
+		LOG_INF("PHY (%d) Link is %s", config->phy_addr, data->state.is_up ? "up" : "down");
+
+		if (data->state.is_up == false) {
```
thanks, added link callback if cb is set.

spectrum70 commented on boards/st/nucleo_h563zi/Kconfig.defconfig:
```diff
@@ -10,6 +10,10 @@ if NETWORKING
 	default y
 
+config PHY_INIT_PRIORITY
+	default 81
```
Well, priority to 81 here is due to the fact that, as far as i have seen, DT structure drives the initialization order. 

The DT structure i chosed is that mdio is child of mac, and phy child of mdio.

Other approach would be that mdio is not a child of mac, but at same level, and to use phy-handle to get the phy, but this method didn't worked, was getting errors compiling using phy-handle = <..>in DT.

Since mdio is part of ETH IP, structure i chosed may be good. Comments are welcome.

decsny replied:
The way you described not working is what I did in nxp_enet , maybe look at that

spectrum70 replied:
@decsny 
I see, it should work, i can eventually change in that way if you prefer, but requires some rework-retest.
If you consider this necessary, more clean, i can go that way.

decsny replied:
I think it would be good to be consistent within zephyr,  unless you think your current way is better

gramsay0 replied:
Have you checked if the initialisation order is only required because both the phy driver and the stm32 eth mac driver (via `HAL_ETH_Init`) initialise the phy?

spectrum70 replied:
@gramsay0 @decsny 

I tested same eth_nxp_enet DT structure, does not work since PHY, with defualt priorities, starts first, and i get
[00:00:00.625,000] <err> adi_adin1100_phy: PHY 0 didn't come out of reset, -116

Issue is the following:

MDIO hw is part of the ETH IP module and actually initialized by eth_stm32_hal.c only, by HAL_ETH_Init().
So PHY for this case must be initialized after ETH.

spectrum70 commented on drivers/ethernet/phy/Kconfig:
```diff
@@ -36,9 +36,9 @@ config PHY_GENERIC_MII
-	depends on MDIO_ADIN2111
+	depends on DT_HAS_ADI_ADIN2111_PHY_ENABLED || DT_HAS_ADI_ADIN1100_PHY_ENABLED
+	depends on ETH_ADIN2111 || ETH_STM32_HAL
+	depends on MDIO_ADIN2111 || MDIO_ST_STM32_HAL
```
fixed, thanks

spectrum70 commented:
Pushed a V3:
- fix erroneous use of memcmp
- add link change calback
- remove unneeded Kconfig dependences

Waiting comments on devicetree part, in case i can move to @decsny nxp_enet structure.

erwango commented on boards/st/nucleo_h723zg/Kconfig.defconfig:
```diff
@@ -10,6 +10,10 @@ if NETWORKING
 config NET_L2_ETHERNET
 	default y
 
+config PHY_INIT_PRIORITY
+	default 81
+	depends on NET_L2_ETHERNET && ETH_DRIVER
+
```
Can't this be moved at SoC level?  Is there a specific reason to make it board specific ?

spectrum70 replied:
Hi @erwango ,
yes, setting it at soc-level looks a better approach, done.

erwango commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -12,6 +12,7 @@ menuconfig ETH_STM32_HAL
 	select NOCACHE_MEMORY if SOC_SERIES_STM32H7X && CPU_CORTEX_M7
 	select HWINFO
 	select ETH_DSA_SUPPORT
+	imply MDIO if SOC_SERIES_STM32H5X || SOC_SERIES_STM32H7X
```
Why using imply here ? Ie is that possible that MDIO would not be available if H5 or H7, and if yes would it be  functional anyway ?

spectrum70 replied:
MDIO hardware and regs are part of ETH IP module, so it's always there when ETH_STM32_HAL is set.

erwango replied:
See "imply" definition here: https://docs.zephyrproject.org/latest/build/kconfig/tips.html#the-imply-statement.
I really think you should use "select" here.

erwango replied:
@spectrum70 Can you check on my feedback on this point ?

spectrum70 replied:
yes., select is what i need. Thanks, fixed.

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -211,8 +217,9 @@ static HAL_StatusTypeDef read_eth_phy_register(ETH_HandleTypeDef *heth,
 {
-#if defined(CONFIG_SOC_SERIES_STM32H7X) ||  defined(CONFIG_SOC_SERIES_STM32H5X) || \
-	defined(CONFIG_ETH_STM32_HAL_API_V2)
+#if defined(CONFIG_SOC_SERIES_STM32H7X) || defined(CONFIG_SOC_SERIES_STM32H5X)
```
if MDIO ?

spectrum70 replied:
Here is the same, still not implemented and tested MDIO support for F7 serie at this stage.

erwango replied:
Since MDIO is only available on H5 and H7 today, then `if MDIO` is the same as `if H5 || H7`.
Benefit is that you won't need to do any modification when MDIO will support F7.

erwango replied:
@spectrum70 Can you check on my feedback on this point ?

spectrum70 replied:
Ok, that's correct. Let me change and test it.

spectrum70 replied:
@erwango  fixed, thanks

spectrum70 replied:
yes, correct, select is what i want.
Fixed, thanks.

erwango reviewed (commented):
Some questions on the STM32 parts, see comments. Otherwise looks good

spectrum70 committed:
soc: st: set soc-level phy priority

Introducing MDIO and PHY support for stm32, phy driver gets
error (-116) if it tries to read phy chip id, since MDIO IP is
part of ETH IP, and eth hw module is still not initialized.

Forcing a priority that allows possibly connected PHY chip to be
detected properly at initial boot.

Signed-off-by: Angelo Dureghello <adureghello@baylibre.com>

spectrum70 commented:
Changes for v4:
- set soc level phy priority

erwango reviewed (dismissed):
Ok for me on STM32 changes.
Though, I'd like we can have some time to test.

erwango commented:
@marwaiehm-st Please test and remove "Block HW test" when done.

marwaiehm-st commented:
Test ok on boards : nucleo_h743zi and  stm32h573i_dk

spectrum70 commented:
@marwaiehm-st thanks ! if you have a chance, you may test still an ethernet FX, to be sure is not damaged from the chage (same ETH driver).

pdgendt commented on drivers/mdio/mdio_stm32_hal.c:
```diff
@@ -0,0 +1,121 @@
+
+	k_sem_give(&dev_data->sem);
+
+	if (ret != HAL_OK)
+		return -EIO;
```
Please add `{}` braces after `if` statements.

```suggestion
	if (ret != HAL_OK) {
		return -EIO;
	}
```

spectrum70 replied:
done

pdgendt commented on drivers/mdio/mdio_stm32_hal.c:
```diff
@@ -0,0 +1,121 @@
+
+	k_sem_give(&dev_data->sem);
+
+	if (ret != HAL_OK)
+		return -EIO;
```
Same here

spectrum70 replied:
thanks, done

spectrum70 commented:
Changes for v5:
- fix missing braces on if statements

spectrum70 commented:
@MaureenHelm @erwango so fixed last issue reported, could you please re-accept, if all ok ? Thanks.

erwango commented:
> @MaureenHelm @erwango so fixed last issue reported, could you please re-accept, if all ok ? Thanks.

I'd like @marwaiehm-st to confirm this has no impact on F7 series first. Otherwise it's fine for me.
This should be done in a couple of days max.

marwaiehm-st commented:
Test ok on board nucleo_f767zi , there is no impact on F7 series first.

aescolar commented:
@spectrum70 can you send a follow up PR adding a line to the release notes? Thanks in advance.
